engine/access/rpc/backend/backend_network.go

```diff
@@ -80,6 +80,54 @@ func (b *backendNetwork) GetLatestProtocolStateSnapshot(_ context.Context) ([]by
 	return data, nil
 }
 
+// GetProtocolStateSnapshotByBlockID returns serializable Snapshot by blockID
+func (b *backendNetwork) GetProtocolStateSnapshotByBlockID(_ context.Context, blockID flow.Identifier) ([]byte, error) {
+	snapshotByBlockId := b.state.AtBlockID(blockID)
+	snapshotHeadByBlockId, err := snapshotByBlockId.Head()
+	if err != nil {
+		return nil, err
+	}
+
+	snapshotByHeight := b.state.AtHeight(snapshotHeadByBlockId.Height)
+	snapshotHeadByHeight, err := snapshotByHeight.Head()
+	if err != nil {
+		return nil, err
+	}
+
+	if snapshotHeadByHeight.ID() != blockID {
+		return nil, status.Errorf(codes.Internal, "failed to retreive snapshot for block ID %v", blockID)
+	}
+
+	validSnapshot, err := b.getValidSnapshot(snapshotByHeight, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	data, err := convert.SnapshotToBytes(validSnapshot)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to convert snapshot to bytes: %v", err)
+	}
+
+	return data, nil
+}
+
+// GetProtocolStateSnapshotByHeight returns serializable Snapshot by block height
+func (b *backendNetwork) GetProtocolStateSnapshotByHeight(_ context.Context, blockHeight uint64) ([]byte, error) {
+	snapshot := b.state.AtHeight(blockHeight)
+
+	validSnapshot, err := b.getValidSnapshot(snapshot, 0)
+	if err != nil {
+		return nil, err
```

Comment:
add a return code, otherwise it's `codes.Unknown`
```suggestion
		return nil, status.Errorf(codes.Internal, "failed to get a valid snapshot: %v", err)
```